src/iterator.rs

```diff
@@ -0,0 +1,298 @@
+    buffer: [u8; READER_BUFFER_SIZE],
+}
+
+/// The contents of an archive, yielded in order from the beginning to the end
+/// of the archive.
+///
+/// Each entry, file or directory, will have a
+/// [`ArchiveContents::StartOfEntry`], zero or more
+/// [`ArchiveContents::DataChunk`], and then a corresponding
+/// [`ArchiveContents::EndOfEntry`] to mark that the entry has been read to
+/// completion.
+pub enum ArchiveContents {
+    /// Marks the start of an entry, either a file or a directory.
+    StartOfEntry(String),
+    /// A chunk of uncompressed data from the entry. Entries may have zero or
+    /// more chunks.
+    DataChunk(Vec<u8>),
+    /// Marks the end of the entry that was started by the previous
+    /// StartOfEntry.
+    EndOfEntry,
+    Err(Error),
+}
+
+/// An iterator over the contents of an archive.
+pub struct ArchiveIterator<R: Read + Seek> {
+    archive_entry: *mut ffi::archive_entry,
+    archive_reader: *mut ffi::archive,
+
+    in_file: bool,
+    closed: bool,
+    error: bool,
+
+    _pipe: Box<HeapReadSeekerPipe<R>>,
+    _utf8_guard: UTF8LocaleGuard,
+}
+
+impl<R: Read + Seek> Iterator for ArchiveIterator<R> {
+    type Item = ArchiveContents;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.error {
+            return None;
+        }
+
+        let next = if self.in_file {
+            unsafe { self.next_data_chunk() }
+        } else {
+            unsafe { self.next_header() }
+        };
+
+        match &next {
+            ArchiveContents::StartOfEntry(_) => {
+                self.in_file = true;
+                Some(next)
+            }
+            ArchiveContents::DataChunk(_) => Some(next),
+            ArchiveContents::EndOfEntry if self.in_file => {
+                self.in_file = false;
+                Some(next)
+            }
+            ArchiveContents::EndOfEntry => None,
+            ArchiveContents::Err(_) => {
+                self.error = true;
+                Some(next)
+            }
+        }
+    }
+}
+
+impl<R: Read + Seek> Drop for ArchiveIterator<R> {
+    fn drop(&mut self) {
+        if !self.closed {
+            unsafe {
+                ffi::archive_read_close(self.archive_reader);
+                ffi::archive_read_free(self.archive_reader);
+            }
+        }
+    }
+}
+
+impl<R: Read + Seek> ArchiveIterator<R> {
+    /// Close the iterator, freeing up the associated resources.
+    ///
+    /// Resources will be freed on drop if this is not called, but any errors
+    /// during freeing on drop will be lost.
+    pub fn close(self) -> Result<()> {
+        unsafe {
+            archive_result(
+                ffi::archive_read_close(self.archive_reader),
+                self.archive_reader,
+            )?;
+            archive_result(
+                ffi::archive_read_free(self.archive_reader),
+                self.archive_reader,
+            )?;
+        }
+        let mut iter = self;
+        iter.closed = true;
+        Ok(())
```

Comment:
This consumes self. It shouldn't otherwise we don't need to set `closed`. Am I wrong?

Reply:
My idea here would be something like:

```diff
diff --git a/src/iterator.rs b/src/iterator.rs
index 91600cd..58a5c9c 100644
--- a/src/iterator.rs
+++ b/src/iterator.rs
@@ -82,12 +82,8 @@ impl<R: Read + Seek> Iterator for ArchiveIterator<R> {
 
 impl<R: Read + Seek> Drop for ArchiveIterator<R> {
     fn drop(&mut self) {
-        if !self.closed {
-            unsafe {
-                ffi::archive_read_close(self.archive_reader);
-                ffi::archive_read_free(self.archive_reader);
-            }
-        }
+        // Drop cannot handle errors during the clean up so we ignore it.
+        let _ = self.close();
     }
 }
 
@@ -96,7 +92,11 @@ impl<R: Read + Seek> ArchiveIterator<R> {
     ///
     /// Resources will be freed on drop if this is not called, but any errors
     /// during freeing on drop will be lost.
-    pub fn close(self) -> Result<()> {
+    pub fn close(&mut self) -> Result<()> {
+        if self.closed {
+            return Ok(());
+        }
+
         unsafe {
             archive_result(
                 ffi::archive_read_close(self.archive_reader),
@@ -107,8 +107,9 @@ impl<R: Read + Seek> ArchiveIterator<R> {
                 self.archive_reader,
             )?;
         }
-        let mut iter = self;
-        iter.closed = true;
+
+        self.closed = true;
+
         Ok(())
     }
 ```

Reply:
I believe the compiler does the right thing and optimizes the result when using it inside the `drop`.

Reply:
I closed the other discussion so we can keep this here, close to the code. You mentioned you prefer the previous ergonomics but I fail to understand what drawback do you see with it?

Reply:
I preferred the older version. The issue now is that you can close it but still continue to try to iterate from it, so we additionally need to check if it's closed inside next(). You can also close it but not drop it, keeping the file open and Rust-side resources allocated.

Though I did notice a subtle bug, it should set `self.closed = true` before freeing the C resources to avoid a potential double free if the first one succeeds.

Reply:
I factored it out into a private .free() method instead.